### DISCUSSION TITLE
can save annotations in separate files?

### DISCUSSION THREAD

**Discussion by AlexShyXie** [sumatrapdfreader/sumatrapdf#4021] [Ideas]
Hello, I would like to inquire whether it is possible to save annotations in SumatraPDF as separate files, such as XML, without modifying the original PDF. This would ensure that the PDF remains clean and pristine.

**Comment by GitHubRulesOK** [comment]
It was possible long time ago (as an experiment) but problems with PDF and other files scaling different ways plus not so easy to manage so no longer possible as separate files.

What you require is one of those products that still do it that way but there are so few left as most no longer maintain such a working model and all are proprietary so not sure if any use XML it is not really suited to the task.

Consider Okular has such a system but some time ago warned it may be depreciated.

**Reply by AlexShyXie** [reply]
Yes, Okular was once supported, but it is 5 or 6 years ago, it stopped being supported. Therefore, I can hardly find such a PDF note-taking tool now, especially on Windows.

**Reply by GitHubRulesOK** [reply]
I suggested okular as it uses xml
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE xbel>
<xbel xmlns:kdepriv="http://www.kde.org/kdepriv" xmlns:mime="http://www.freedesktop.org/standards/shared-mime-info" dbusName="okular" xmlns:bookmark="http://www.freedesktop.org/standards/desktop-bookmarks">
 <folder href="file:///C:/Users/lez/Downloads/123 (cleaner).pdf">
  <title>C:/Users/lez/Downloads/123 (cleaner).pdf</title>
  <bookmark href="file:///C:/Users/lez/Downloads/123%20(cleaner).pdf#0;C2:0.499432:0.208901:1">
   <title>pypdf</title>
   <info>
    <metadata owner="http://freedesktop.org">
     <bookmark:icon name=""/>
    </metadata>
   </info>
  </bookmark>
 </folder>
</xbel>
```
my running version is 2019 because that's  stable on win32 but the latest version as a 64bit build is [Build #1638 (12-Jan-2024)](https://binary-factory.kde.org/job/Okular_Release_win64/1638/)

**Reply by AlexShyXie** [reply]
> I suggested okular as it uses xml
> ```
> <?xml version="1.0" encoding="UTF-8"?>
> <!DOCTYPE xbel>
> <xbel xmlns:kdepriv="http://www.kde.org/kdepriv" xmlns:mime="http://www.freedesktop.org/standards/shared-mime-info" dbusName="okular" xmlns:bookmark="http://www.freedesktop.org/standards/desktop-bookmarks">
>  <folder href="file:///C:/Users/lez/Downloads/123 (cleaner).pdf">
>   <title>C:/Users/lez/Downloads/123 (cleaner).pdf</title>
>   <bookmark href="file:///C:/Users/lez/Downloads/123%20(cleaner).pdf#0;C2:0.499432:0.208901:1">
>    <title>pypdf</title>
>    <info>
>     <metadata owner="http://freedesktop.org">
>      <bookmark:icon name=""/>
>     </metadata>
>    </info>
>   </bookmark>
>  </folder>
> </xbel>
> ```
> my running version is 2019 because that's  stable on win32 but the latest version as a 64bit build is [Build #1638 (12-Jan-2024)](https://binary-factory.kde.org/job/Okular_Release_win64/1638/)

Thanks, I would download the version you mentioned to have a try.

**Reply by AlexShyXie** [reply]
Sir, I did not find the format 'xml' in `save as` or `export`, the `okular` format is also Embed the total pdf file. And I can't download the 2019 version from official website:
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/49368033/818075f3-53bd-4707-bc50-421efef9cdd0)

**Reply by AlexShyXie** [reply]
And other say it was 0.15 that the last version support saving annotation seperately, so I want to find a new close substitute:

![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/49368033/798ffbf2-081d-4b4b-b987-8835f10558ad)

**Reply by GitHubRulesOK** [reply]
I have several 32bit versions
20v23 (my variant does not include the non PDF loading so does not include okular format)
20v19 (my version is odd as non installed and does load many formats but not build the okular files, just the bookmarks)
20v11 (this oldest version works well as a windows standalone copy and works with okular files) 

Okular files are a zip archive that contain the XML and say an image with markup overlay the content list showing it is a bmp
```
<OkularArchive>
<Files>
<DocumentFileName>4500.bmp</DocumentFileName>
<MetadataFileName>metadata.xml</MetadataFileName>
</Files>
</OkularArchive>
```
the metadata.xml file may then have the overlay
```
<documentInfo>
<pageList>
<page number="0">
<annotationList>
<annotation type="6">
<base creationDate="2024-01-14T02:50:38" uniqueName="okular-0-1" author=" " modifyDate="2024-01-14T02:50:38" color="#00ff00">
<boundary l="0.0319361" r="0.484032" b="0.606742" t="0.137328"/>
<penStyle width="2" ycr="0" style="1" spaces="0" marks="3" xcr="0"/>
</base>
<ink>
<path>
<point x="0.291417" y="0.167291"/>
<point x="0.288423" y="0.167291"/>
<point x="0.277445" y="0.167291"/>
<point x="0.264471" y="0.167291"/>
<point x="0.253493" y="0.167291"/>
<point x="0.245509" y="0.167291"/>
<point x="0.238523" y="0.167291"/>

many more

<point x="0.258483" y="0.139825"/>
<point x="0.257485" y="0.139825"/>
<point x="0.255489" y="0.139825"/>
</path>
</ink>
</annotation>
</annotationList>
</page>
</pageList>
</documentInfo>
```

So if you can save as okular then simply rename that file to zip or use Windows TAR to extract the content.
SumatraPDF would show the BMP inside the zip but NOT overlay the XML

ALL Okular versions will read the okular file format but my V23 does not have the necessary library to show the contents.

**Reply by AlexShyXie** [reply]
Thank you very much, I get it.

**Reply by AlexShyXie** [reply]
Sir, I use v23, `*.okular` can indeed be opened with zip, and there are also two xml files inside, this is a huge breakthrough. But I found that `metadata.xml` does not contain any annotations information. Then, I extracted the PDF file from it and found that the annotations were still on the PDF. So currently, even if Okular is saved as `.okular` format, the annotations are still written on the PDF. 
Aren’t you like this?
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/49368033/dcb64454-ce3f-4ea8-be88-ae8fbaa18803)

**Reply by GitHubRulesOK** [reply]
It can only save NEW annotations by save as Okular after they are added. Which was the case with earlier versions , but I cannot say if that is still the case as my non working v23 does not add any. (Rumours were they would change the method but never clear when)

It was the same with SumatraPDF.SMX files where if previous annotations were appended to the PDF they were not separable after that. Thus with the new system of Adobe incremental appending, there too can be problems but, more compatible types now with other PDF viewers than before. (Allows for delete and replace)

To extract older annotations and convert to XML you would need to read/output them, then delete them, then add them again using the "SideCar" file. For SumatraPDF the last SMX reading version was 3.2 not 3.3+

**Reply by AlexShyXie** [reply]
Thank you very much!

**Reply by AlexShyXie** [reply]
I am sad, it seems the newest version of Okular no longer save annotations in metadata.xml, I tried a lot but metadata.xml did not change.
I gave up.
